File: bsp4kt/src/com/jetbrains/bsp/bsp4kt/CargoBuildServer.kt
package com.jetbrains.bsp.bsp4kt

import com.jetbrains.jsonrpc4kt.services.JsonNotification
import com.jetbrains.jsonrpc4kt.services.JsonRequest
import kotlinx.serialization.SerialName
import java.util.concurrent.CompletableFuture

interface CargoBuildServer {
  @JsonRequest("workspace/cargoFeaturesState")
  fun cargoFeaturesState(): CompletableFuture<CargoFeaturesStateResult>

  @JsonRequest("workspace/enableCargoFeatures")
  fun enableCargoFeatures(params: EnableCargoFeaturesParams): CompletableFuture<Unit>

  @JsonRequest("workspace/disableCargoFeatures")
  fun disableCargoFeatures(params: DisableCargoFeaturesParams): CompletableFuture<Unit>

}

File: bsp4rs/src/com/jetbrains/bsp/bsp4rs/cargo/cargo_build_server.rs
use crate::*;

#[derive(Debug)]
pub enum CargoFeaturesState {}

impl Request for CargoFeaturesState {
    type Params = ();
    type Result = CargoFeaturesStateResult;
    const METHOD: &'static str = "workspace/cargoFeaturesState";
}

#[derive(Debug)]
pub enum EnableCargoFeatures {}

impl Request for EnableCargoFeatures {
    type Params = EnableCargoFeaturesParams;
    type Result = ();
    const METHOD: &'static str = "workspace/enableCargoFeatures";
}

#[derive(Debug)]
pub enum DisableCargoFeatures {}

impl Request for DisableCargoFeatures {
    type Params = DisableCargoFeaturesParams;
    type Result = ();
    const METHOD: &'static str = "workspace/disableCargoFeatures";
}
